### PR TITLE
NEW: Monitor cost function for convergence

### PR DESCRIPTION
**Pull request by carterbox**
## Purpose

Monitor the cost function for convergence to allow for early termination or other convergence related features.

## Approach

Use numpy polyfit to compute the slope of the cost function across some window of the latest epochs. If the slope is no longer decreasing, then the algorithm has converged!

We use an average cost function value from all mini-batches if there is more than one. Also, the cost from all mini-batches is now reported in the algorithm_options.costs parameter.

### Submitter
- [x] Write a helpfully descriptive pull request title.
- [x] Organize changes into logically grouped commits with descriptive commit messages.
- [x] Document all new functions.
- [x] Click 'details' on the readthedocs check to view the updated docs.
- [x] Write tests for new functions or explain why they are not needed.
- [x] Address any complaints from pep8speaks.

### Reviewer
- [x] Actually read all of the code.
- [x] Run the new code yourself; the included tests should make this easy.
- [x] Write a summary of the changes as you understand them.
- [x] Thank the submitter.

### CODE REVIEW

File: src/tike/opt.py [hunk 1]
```diff
@@ -16,6 +16,27 @@
 randomizer = np.random.default_rng()
 
 
+def is_converged(algorithm_options):
```
Comment (carterbox) [inline src/tike/opt.py:19]: Rename to relative or absolute.

File: src/tike/opt.py [hunk 2]
```diff
@@ -16,6 +16,27 @@
...
+    window cost values (average cost values if mini-batch). If this slope is
+    non-negative, return True else return False.
+    """
+    window = algorithm_options.convergence_window
+    if (window >= 2 and len(algorithm_options.costs) >= window
+            and len(algorithm_options.costs) % window // 2 == 0):
+        m = np.array(algorithm_options.costs[-window:])
+        m = np.reshape(m, (len(m), -1))
+        m = np.mean(m, axis=1)
+        p = np.polyfit(x=range(window), y=m, deg=1, full=False, cov=False)
```
Comment (carterbox) [inline src/tike/opt.py:32]: Try loglog plot.

**Comment by carterbox** [comment]
Figure_4
Figure_7

I don't think applying a loglog transformation to the cost values before computing the slope is helpful. Because it just makes the flat regions flatter?

**Comment by stevehenke** [comment]
I agree with your conclusion. This convergence behavior is something that we should be prepared to explain. In this case, it's probably best to go ahead with the simple solution that you've implemented.

**Review by stevehenke — APPROVED** [review]
This PR adds a convergence monitor to tike. The code is straightforward and tests are passing. Thanks, Daniel!

**Pull request merged**